tb_plugin/test/test_tensorboard_end2end.py

```diff
@@ -28,51 +37,49 @@ def test_tensorboard_end2end(self):
             link_prefix + 'kernel?run={}&worker=worker0&view=Kernel&group_by=Kernel'
         ]
 
-        tb = Popen(['tensorboard', '--logdir='+test_folder, '--port='+str(port)])
-
-        timeout = 60
+        retry_times = 60
         while True:
             try:
                 socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
                 print('tensorboard start successfully')
                 break
             except socket.error:
                 time.sleep(2)
-                timeout -= 1
-                if timeout < 0:
-                    tb.kill()
-                    raise RuntimeError("tensorboard start timeout")
+                retry_times -= 1
+                if retry_times < 0:
+                    self.fail("tensorboard start timeout")
                 continue
 
-        timeout = 60
+        retry_times = 60
 
         while True:
             try:
                 response = urllib.request.urlopen(run_link)
                 if response.read()==expected_runs:
                     break
+                if retry_times % 10 == 0:
+                    print("receive mismatched data, retrying", response.read())
                 time.sleep(2)
-                timeout -= 1
-                if timeout<0:
-                    tb.kill()
-                    raise RuntimeError("Load run timeout")
-            except Exception:
-                continue
+                retry_times -= 1
+                if retry_times<0:
+                    self.fail("Load run timeout")
+            except Exception as e:
+                if retry_times > 0:
+                    continue
+                else:
+                    print(e)
+                    self.fail("exception happens {}".format(e))
```

Comment:
this is s test code. We'd like to print the exception and fail the test here.

Reply:
`raise e` will fail the test automatically (uncaught exceptions will fail unit tests). And of course in the test results, you'll see the full stack trace. But when you convert it to a string by inserting it into that "exception happens" string, you'll *only* keep the message and not the stack trace. I'd recommend keeping things simple here and just raising the exception directly, rather than printing it and then failing (which also means you're outputting the error message twice).

Reply:
> I still think it would be better to simply `import` the original gfile, and then add your own methods (or override when necessary the original ones). Maybe others can comment here as well with their thoughts - but making your own version means that you need to manually pull in new fixes from the original gfile if there are optimizations/patches that go into it.

We need change lot of logics in gfile.py and simply import it is not enough. For example, there is no download file support.

Reply:
Did you intend to put that comment in a different section? Raising the exception directly is still better and your comment doesn't seem related to this.

Reply:
sorry, I put wrong comment here. will change it in a separated PR.